### ISSUE TITLE
Getting "No matching state found in storage" error

### ISSUE THREAD

**Issue by YogiDhingani27** [authts/oidc-client-ts#1487]
When I'm trying to visit a url which has a code as a query param then authentication is failing and giving this error.

Other urls are working fine for ex: http://localhost:3004/integration

Not working url: 

http://localhost:3000/integration?state=%7B%22uuid%22:%20%1234%22,%20%22platform%22:%20%22google%22%7D&code=1233434&scope=email%20profile%20https://www.googleapis.com/auth/userinfo.profile%20https://www.googleapis.com/auth/userinfo.email%20https://www.googleapis.com/auth/drive.readonly%20openid&authuser=0&hd=werwer.com&prompt=consent

![image](https://github.com/authts/react-oidc-context/assets/94128664/472a749a-a625-4be2-8423-57500ff95c0d)

Note: Excluding url having code as a query parameter other urls are working fine.

I want a code query parameter in my url for OAuth related authentication.

**Comment by YogiDhingani27** [comment]
UserManager Configuration:

`export const userManager = new UserManager({ authority: authority, client_id: clientId, redirect_uri: window.location.origin + window.location.pathname, post_logout_redirect_uri: window.location.origin, monitorSession: true });`

I'm using keycloak for authentication.

**Comment by pamapa** [comment]
You will need to debug/log your specific application issue. Using https://authts.github.io/oidc-client-ts/#md:logging.

When the authentication process starts this library add a  state to the session storage, this state later when the callback is received must be found again.

Store: https://github.com/authts/oidc-client-ts/blob/main/src/OidcClient.ts#L145
Read: https://github.com/authts/oidc-client-ts/blob/main/src/OidcClient.ts#L159

**Comment by YogiDhingani27** [comment]
@pamapa  I'm using keycloak for authentication there is nothing i can debug. Initiallly when i hit the url, keycloak will automatically append the code into url for authentication purpose. So i think oidc is considering the wrong code for authentication as above mentioned. i have a two code keys.

And above mentioned url is working if i'm using react-keycloak/web.

**Comment by vvikene** [comment]
Are you doing a round trip every time you visit a page? 
You are using whatever page visited as the redirect callback url.. how do you check if it's a redirect callback or not?

**Comment by YogiDhingani27** [comment]
> Are you doing a round trip every time you visit a page? You are using whatever page visited as the redirect callback url.. how do you check if it's a redirect callback or not?

I'm not checking the redirect url. oidc is giving the error when you have code query param in url.

**Comment by YogiDhingani27** [comment]
Here is the video of the issue which you can refer:

[keycloak-issue.webm](https://github.com/user-attachments/assets/45c7d813-00ff-44c3-84bf-ad94464fda65)

**Comment by YogiDhingani27** [comment]
@pamapa 

If user includes state and code as a query params in url then it is giving this error. 

Can you please help me how can i solve this issue?

**Comment by Hatko** [comment]
Having the same issue

**Comment by apavelm** [comment]
same issue here

**Comment by atomovski** [comment]
@YogiDhingani27 / @Hatko / @apavelm - are you guys using [`react-oidc-context`](https://github.com/authts/react-oidc-context)? 

I ran into a similar issue and the root cause was a function that checks if the `code` search param exists [here](https://github.com/authts/react-oidc-context/blob/d13817707f4d77ab32ef7fba9cb3e97facedb91f/src/utils.ts#L4) and then proceeds to call `userManager.signinCallback` as per their documentation [here](https://github.com/authts/react-oidc-context/blob/d13817707f4d77ab32ef7fba9cb3e97facedb91f/src/AuthProvider.tsx#L39)

TL;DR; a simple fix is to add `skipSigninCallback` like this `<OidcAuthProvider {...oidcConfig} skipSigninCallback>`

**Comment by zaidmasri** [comment]
> [@YogiDhingani27](https://github.com/YogiDhingani27) / [@Hatko](https://github.com/Hatko) / [@apavelm](https://github.com/apavelm) - are you guys using [`react-oidc-context`](https://github.com/authts/react-oidc-context)?
> 
> I ran into a similar issue and the root cause was a function that checks if the `code` search param exists [here](https://github.com/authts/react-oidc-context/blob/d13817707f4d77ab32ef7fba9cb3e97facedb91f/src/utils.ts#L4) and then proceeds to call `userManager.signinCallback` as per their documentation [here](https://github.com/authts/react-oidc-context/blob/d13817707f4d77ab32ef7fba9cb3e97facedb91f/src/AuthProvider.tsx#L39)
> 
> TL;DR; a simple fix is to add `skipSigninCallback` like this `<OidcAuthProvider {...oidcConfig} skipSigninCallback={window.location.pathname === "/example/path"}>`

I tried this but it caused in infinite loop. I'm using `react-oidc-context`. 

This is my redirect url: 

http://localhost:8891/property?state=bfab493caac54c73982ea3f1d83c8901&session_state=a477bade-1c85-45ae-a147-42a119bc88de&iss=https%3A%2F%2Fkeycloak.localhost%2Frealms%2Frms&code=1d51280c-8189-4170-bfc1-7bd50c0fbba3.a477bade-1c85-45ae-a147-42a119bc88de.fd8fc2e8-6593-40c9-9467-d95815a6a987

**Comment by apavelm** [comment]
@atomovski No, I'm using oidc-client-ts directly. Previously, I tried to use different Vue* libraries, but each has some issues, so I decided to use oidc-client-ts directly and build own plugin.

BTW, I solved the issue. I noticed that sometimes calls for processing callbacks called twice. I wrapped UserManager to be used from  a singelton class, and revised all async/awaits. Now it works

**Comment by asbjornu** [comment]
I'm having this issue trying to invoke `UserManager.signinCallback()` on the server. Seems like this is only made to work client-side in the browser.

**Comment by zach-betz-hln** [comment]
Hi @YogiDhingani27 (and others) - am late to the party but will ask anyway - are you able to reproduce this issue in the [sample](https://github.com/authts/sample-keycloak-react-oidc-context) repo? (which also uses Keycloak as the identity provider)
